man/man3/str.3
.\" 
.\" Copyright (c) 2011, Gordon D. Carrie. All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 
.\"     * Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"     * Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
.\" TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" Please address questions and feedback to dev0@trekix.net
.\" 
.\" $Revision: 1.19 $ $Date: 2011/11/28 16:11:23 $
.\"
.TH str 3 "String manipulation functions"
.SH NAME
Str_Esc, Str_Words \- manipulate strings
.SH SYNOPSIS
.nf
\fB#include "str.h"\fP
\fBchar *Str_Esc(char *\fP\fIs\fP\fB);\fP
\fBchar\fP ** \fBStr_Words\fP(\fBchar\fP *\fIln\fP, \fBchar\fP **\fIargv\fP, \fBint\fP *\fIargc\fP);
\fBchar\fP * \fBStr_Append\fP(\fBchar\fP *\fIdest\fP, \fBsize_t\fP *\fIl\fP, \fBsize_t\fP *\fIlx\fP, \fBchar\fP *\fIsrc\fP, \fBsize_t\fP \fIn\fP);
\fBint\fP \fBStr_GetLn\fP(\fBFILE\fP *\fIin\fP, \fBchar\fP \fIeol\fP, \fBchar\fP **\fIln\fP, \fBint\fP *\fIl_max\fP);
.fi
.SH DESCRIPTION
\fBStr_Esc\fP replaces escape sequences in \fIs\fP with their corresponding ASCII
characters.
The following escape sequences are recognized.
.ta 8m
.br
\fB\\a\fP	becomes a single <bell> character.
.br
\fB\\b\fP	becomes a single <backspace> character.
.br
\fB\\f\fP	becomes a single <form-feed> character.
.br
\fB\\n\fP	becomes a single <new-line> character.
.br
\fB\\r\fP	becomes a single <carriage return> character.
.br
\fB\\t\fP	becomes a single <tab> character.
.br
\fB\\v\fP	becomes a single <vertical tab> character.
.br
\fB\\'\fP	becomes a single <single quote> character.
.br
\fB\\\\ \fP	becomes a single backslash character.
.br
.in +8m
.ti -8m
\fB\\0\fP\fInum\fP	becomes the character whose ASCII value is \fInum\fP, where \fInum\fP is an integer given as 1, 2, or 3 octal digits.
.in -8m
The (possibly modified) string is returned.

\fBStr_Words\fP splits \fBnul\fP terminated string \fIln\fP into words.
A word is a contiguous sequence of non-whitespace characters,
or a contiguous sequence of any characters bounded by a matching pair of quotes
(\fB"\fP or \fB'\fP).
If modifies \fIln\fP in place as follows.
Runs of whitespace between words are replaced with single \fBnul\fP characters,
and quotes bounding words are deleted.
Return value is a \fBNULL\fP terminated array of pointers to the start of each
word in \fIln\fP.
The number of words found is placed at \fIargc\fP. 

If \fIargv\fP is \fBNULL\fP upon entry, a new allocation is made and its address
is returned.  Memory for the returned array should eventually be freed with a call
to \fBfree\fP.

If \fIargv\fP is not \fBNULL\fP, it should point to an allocation made with
\fBmalloc\fP, \fBcalloc\fP, or \fBrealloc\fP with storage for
*\fIargc\fP\ \fB\ +\ 1\fP strings. \fBStr_Words\fP may reallocate \fIargv\fP
with \fBrealloc\fP and update *\fIargc\fP if it finds more than *\fIargc\fP words.
The (possibly new) allocation is returned.

If something goes wrong, \fBNULL\fP is returned, \fIargv\fP is freed,
\fI*argc\fP is set to \fB-1\fP, and an error message is printed to standard error.

\fBStr_Append\fP copies \fIn\fP characters from \fIsrc\fP to \fIdest\ +\ *l\fP and
appends a \fBnul\fP.  At entry, \fIdest\fP should point to storage for *\fIlx\fP
characters allocated by \fBmalloc\fP, \fBcalloc\fP, or \fBrealloc\fP.  If space is
insufficient, it adjusts the allocation with \fBrealloc\fP and updates \fIlx\fP.
To avoid excessive allocation adjustments, it usually allocates extra space at
\fIdest\fP.
Return value is \fIdest\fP or a new allocation.  Upon return, *\fIl\fP indicates
the number of characters at \fIdest\fP, excluding the terminating \fBnul\fP.  If
something goes wrong, \fBStr_Append\fP returns \fBNULL\fP.  Callers that provide
a non-\fBNULL\fP \fIdest\fP should not assign the return value to \fIdest\fP
until the caller has verified a valid return value.  (This is similar to using
realloc.) An error message will be printed to standard error.

\fBStr_GetLn\fP reads characters from input stream \fIin\fP until it encounters
end of line character \fIeol\fP or end of file.  The resulting string is placed
at *\fIln\fP with a \fBnul\fP terminator instead of the end of line character, and
the allocation size, including the terminating \fBnul\fP, is stored at \fIl_max\fP.
If, upon entry, *\fIln\fP is \fBNULL\fP, the string will be returned in a new
allocation.  The memory should eventually be freed with a call to \fBfree\fP.
If *\fIln\fP is not \fBNULL\fP, it is assumed to point to storage for
*\fIl_max\fP\ +\ \fB1\fP characters, allocated with \fBmalloc\fP, \fBcalloc\fP,
or \fBrealloc\fP.  \fBStr_GetLn\fP will store characters at the given memory,
clobbering any previous contents.  If it encounters more than
*\fIl_max\fP\ -\ \fB1\fP characters, it will increase the allocation with
\fBrealloc\fP and update *\fIl_max\fP.  If \fBStr_GetLn\fP encounters \fIeol\fP,
it returns \fB1\fP.  If it encounters end of file, if returns \fBEOF\fP.
If something goes wrong, \fBStr_GetLn\fP, frees *\fIln\fP returns \fB0\fP, and
prints an error message to standard error.
.SH AUTHOR
Gordon Carrie (dev0@trekix.net)


src/str.c
/*
   -	str.c --
   -		This file defines string manipulation
   -		functions.  See str (3).
   -	
   .	Copyright (c) 2011, Gordon D. Carrie. All rights reserved.
   .	
   .	Redistribution and use in source and binary forms, with or without
   .	modification, are permitted provided that the following conditions
   .	are met:
   .	
   .	    * Redistributions of source code must retain the above copyright
   .	    notice, this list of conditions and the following disclaimer.
   .
   .	    * Redistributions in binary form must reproduce the above copyright
   .	    notice, this list of conditions and the following disclaimer in the
   .	    documentation and/or other materials provided with the distribution.
   .	
   .	THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   .	"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   .	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   .	A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   .	HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   .	SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
   .	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   .	PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   .	LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   .	NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   .	SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
   .
   .	Please send feedback to dev0@trekix.net
   .
   .	$Revision: 1.23 $ $Date: 2011/11/28 16:09:55 $
 */

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <ctype.h>
#include <limits.h>
#include "str.h"

/* Largest possible number of elements in array of strings */
static int mx = INT_MAX / (int)sizeof(char *);

char *Str_Esc(char *str)
{
    char *p, *q, t;
    int n;

    for (p = q = str; *q; p++, q++) {
	if (*q == '\\') {
	    switch (*++q) {
		/* Replace escape sequence with associated character */
		case 'a':
		    *p = '\a';
		    break;
		case 'b':
		    *p = '\b';
		    break;
		case 'f':
		    *p = '\f';
		    break;
		case 'n':
		    *p = '\n';
		    break;
		case 'r':
		    *p = '\r';
		    break;
		case 't':
		    *p = '\t';
		    break;
		case 'v':
		    *p = '\v';
		    break;
		case '\'':
		    *p = '\'';
		    break;
		case '\\':
		    *p = '\\';
		    break;
		case '0':
		    /* Escape sequence is a sequence of octal digits. */
		    n = (int)strspn(++q, "01234567");
		    t = *(q + n);
		    *(q + n) = '\0';
		    *p = (char)strtoul(q, &q, 8);
		    *q-- = t;
		    break;
		default:
		    *p = *q;
	    }
	} else {
	    *p = *q;
	}
    }
    *p = '\0';

    return str;
}

char ** Str_Words(char *ln, char **argv, int *argc)
{
    char **v = NULL;		/* Array of words from ln. */
    char **t;			/* Temporary */
    int cx;			/* Number of words that can be stored at v */
    int cx2;			/* New value for cx when reallocating */
    int c;			/* Number of words in ln */
    char *p, *q;		/* Pointers into ln */
    int inwd;			/* If true, p points into a word */
    size_t sz;			/* Temporary */

    if (argv) {
	v = argv;
	cx = *argc;
	if (cx < 2) {
	    if ( !(t = (char **)realloc(v, 2 * sizeof(char *))) ) {	/* re */
		free(v);
		*argc = -1;
		fprintf(stderr, "Could not allocate word array.\n");
		return NULL;
	    }
	    v = t;
	    cx = 2;
	}
    } else {
	cx = 2;
	if ( !(v = (char **)calloc((size_t)cx, sizeof(char *))) ) {	/* new */
	    *argc = -1;
	    fprintf(stderr, "Could not allocate word array.\n");
	    return NULL;
	}
    }
    for (p = q = ln, inwd = 0, c = 0; *p; p++) {
	if ( isspace(*p) ) {
	    if ( inwd ) {
		*q++ = '\0';
	    }
	    inwd = 0;
	} else {
	    if ( !inwd ) {
		/* Have found start of a new word */
		if (c + 1 > cx) {
		    cx2 = 3 * cx / 2 + 4;
		    sz = (size_t)cx2 * sizeof(char *);
		    if (cx2 > mx || !(t = (char **)realloc(v, sz)) ) {
			free(v);
			*argc = -1;
			fprintf(stderr, "Could not allocate word array.\n");
			return NULL;
		    }
		    v = t;
		    cx = cx2;
		}
		v[c++] = q;
		inwd = 1;
	    }

	    if ( *p == '"' || *p == '\'' ) {
		/* Append run of characters bounded by quotes */
		char *e = strchr(p + 1, *p);

		if ( !e ) {
		    fprintf(stderr, "Unbalanced quote.\n");
		    free(v);
		    *argc = -1;
		    return NULL;
		}
		strncpy(q, p + 1, e - p - 1);
		q += e - p - 1;
		p = e;
	    } else {
		/* Append character */
		*q++ = *p;
	    }
	}
    }
    *q = '\0';
    *argc = c;
    sz = (size_t)(c + 1) * sizeof(char *);
    if ( (c + 1) > mx || !(t = (char **)realloc(v, sz)) ) {
	free(v);
	*argc = -1;
	fprintf(stderr, "Could not allocate word array.\n");
	return NULL;
    }
    v = t;
    v[c] = NULL;
    return v;
}

char * Str_Append(char *dest, size_t *l, size_t *lx, char *src, size_t n)
{
    char *t;
    size_t lx2;

    lx2 = *l + n + 1;
    if ( *lx < lx2 ) {
	size_t tx = *lx;

	while (tx < lx2) {
	    tx = (tx * 3) / 2 + 4;
	}
	if ( !(t = realloc(dest, tx)) ) {
	    fprintf(stderr, "Could not grow string for appending.\n");
	    return NULL;
	}
	dest = t;
	*lx = tx;
    }
    strncpy(dest + *l, src, n);
    *l += n;
    return dest;
}

int Str_GetLn(FILE *in, char eol, char **ln, int *l_max)
{
    int i;			/* Input character */
    char c;			/* Input character */
    char *t, *t1;		/* Input line */
    size_t n;			/* Number of characters in ln */
    size_t nx;			/* Temporarily hold value for *l_max */

    if ( !*ln ) {
	nx = 4;
	if ( !(t = (char *)malloc((size_t)nx)) ) {
	    fprintf(stderr, "Could not allocate memory for line.\n");
	    return 0;
	}
    } else {
	t = *ln;
	nx = *l_max;
    }
    n = 0;
    while ( (i = fgetc(in)) != EOF && (i != eol) ) {
	c = i;
	if ( !(t1 = Str_Append(t, &n, &nx, &c, (size_t)1)) ) {
	    free(t);
	    fprintf(stderr, "Could not append input character to string.\n");
	    return 0;
	}
	t = t1;
    }
    if ( !(*ln = (char *)realloc(t, (size_t)(n + 1))) ) {
	free(t);
	*ln = NULL;
	*l_max = 0;
	fprintf(stderr, "Could not reallocate memory for line.\n");
	return 0;
    }
    *(*ln + n) = '\0';
    *l_max = n + 1;
    if ( ferror(in) ) {
	perror(NULL);
	return 0;
    } else if ( feof(in) ) {
	return EOF;
    } else {
	return 1;
    }
}
